src/goto-programs/string_instrumentation.cpp

```diff
@@ -610,6 +649,199 @@ void string_instrumentationt::do_strncmp(
+                                return_array_length.type}};
+  const auto apply_associate_length_to_array = function_application_exprt{
+    cprover_associate_length_to_array_func_symbol,
+    {nondet_array_expr, refined_string_expr.length()},
+    index_type()};
+  const auto return_length_expr = return_array_length.symbol_expr();
+  program.add(goto_programt::make_decl(return_length_expr));
+  program.add(goto_programt::make_assignment(
+    return_length_expr, apply_associate_length_to_array));
+
+  // cprover_associate_array_to_pointer_func(nondet_infinite_array_pointer,
+  //                                         src);
+  const address_of_exprt array_pointer(
+    index_exprt(nondet_array_expr, from_integer(0, index_type())));
+  program.add(goto_programt::make_assignment(array_pointer, char_ptr));
+  const symbolt return_sym_pointer =
+    new_aux_symbol("return_array_pointer", index_type(), symbol_table);
+  const auto cprover_associate_array_to_pointer_func_symbol =
+    symbol_exprt{ID_cprover_associate_array_to_pointer_func,
+                 mathematical_function_typet{
+                   {nondet_array_expr.type(), array_pointer.type()},
+                   return_sym_pointer.type}};
+  const auto apply_associate_pointer_to_array =
+    function_application_exprt{cprover_associate_array_to_pointer_func_symbol,
+                               {nondet_array_expr, char_ptr}};
+  const auto return_pointer_expr = return_sym_pointer.symbol_expr();
+  program.add(goto_programt::make_decl(return_pointer_expr));
+  program.add(goto_programt::make_assignment(
+    return_pointer_expr, apply_associate_pointer_to_array));
+
+  return refined_string_exprt{refined_string_expr.length(), char_ptr};
+}
+
+// Implement strchr in terms of indexOf for the purposes of using the
+// string-solver.
+//
+// char * strchr(char *src, int c) {
+//   int string_length = nondet();
+//   assume (string_length < =max_nondet_string_length);
+//   char *string_content = src;
+//   char (*nondet_infinite_array_pointer)[\infty];
+//   nondet_infinite_array_pointer = string_content;
+//
+//   cprover_associate_length_to_array_func(nondet_infinite_array_pointer,
+//                                          string_length);
+//   cprover_associate_array_to_pointer_func(nondet_infinite_array_pointer,
+//                                           src);
+//   struct refined_string_struct {
+//     int length = string_length;
+//     char *content = src;
+//   } refined_string;
+//
+//   maybe_return = cprover_string_index_of_func(refined_string, c);
+//   if (maybe_return >= 0)
+//     return src + maybe_return;
+//   else
+//     return NULL:
+// }
+void string_instrumentationt::do_strchr_via_c_index_of(
+  goto_functiont &strchr_function)
+{
+  //   struct refined_string_struct {
+  //     int length = string_length;
+  //     char *content = src;
+  //   } refined_string;
+  const auto &params = to_code_type(strchr_function.type).parameters();
```

Comment:
:bulb: Feels like some invariants on the parameters might save some tricky debugging